### PR TITLE
fix(layers): don't drop call-context kwargs from the recorded Node when conversion is skipped

### LINKED ISSUE #23294: tree walks still run on single-tensor calls whose structure is already known
Once `Layer.__call__`'s CallSpec fast path has established that a call is a single leaf tensor (at most `training=<bool>`), later steps in the same call still re-derive that already-known structural fact using generic tree machinery. Specifically: the mask pipeline's `tree.map_structure(get_keras_mask, ...)` calls, an `any_symbolic_tensors` re-walk over the inputs, and a general-path `tree.flatten` inside `_set_mask_metadata`. Each of these re-walks a structure that the fast path already proved is a single leaf, so for the (very common) single-tensor-input case they're pure overhead paid on every call.

The fix hoists a single-tensor-call predicate and adds leaf fast branches ahead of the tree-based mask handling, so single-tensor calls skip the redundant walks entirely; the general nested-structure paths are untouched for multi-input/multi-output layers. Fixed by #23300.

Part of the per-call Python-dispatch overhead series in #22561.

### PR DESCRIPTION
**Pull request by pctablet505**
Closes #23294

## Summary

`Layer.__call__` binds `original_kwargs = kwargs` and records that same object into the `Node` at step 8. When the input-conversion branch runs it rebinds `kwargs` to a fresh dict, which is what has been protecting the recorded value. When conversion is skipped, `kwargs` stays bound to the caller's dict — and steps 5, 6 and `CallSpec` all mutate it. `CallSpec` pops call-context arguments that `call()` does not accept, so `training` disappears from the record.

This is reachable on master today. Any layer with `_convert_input_args = False` — which is what `Functional` sets — skips conversion:

```python
class Wrapper(layers.Layer):
    def __init__(self, **kw):
        super().__init__(**kw)
        self._convert_input_args = False
        self.bn = layers.BatchNormalization(momentum=0.0)
    def call(self, x):                      # no `training` parameter
        return self.bn(x)

inp = keras.Input(shape=(4,), batch_size=4)
model = keras.Model(inp, Wrapper()(inp, training=True))
model(np.arange(16, dtype="float32").reshape(4, 4))
```

| | recorded Node kwargs | output mean abs | mode |
|---|---|--:|---|
| master | `{}` | 7.4963 | inference |
| this PR | `{'training': True}` | 0.8944 | training |

The model was built with `training=True`, so replaying it should run the wrapped `BatchNormalization` in training mode. On master it silently runs in inference mode instead. Nothing errors and nothing warns.

The fix is to copy `kwargs` whenever the conversion branch is skipped, so the dict recorded in the `Node` is the one the caller passed.

## The conversion gate

The same change lets a lone `training=<bool-or-None>` skip input conversion. `fit()`, `evaluate()` and `predict()` always pass `training=` explicitly, so a non-empty `kwargs` was not by itself a reason to walk the structure — that shape holds no array-like leaf for `maybe_convert` to act on.

Be aware this is worth little on its own: `Functional` sets `_convert_input_args = False` and so never reached the gate anyway, meaning **Functional models see no speed-up at all**. Subclassed models go from 10 `map_structure` walks per call to 6, and a bare `BatchNormalization` from 4 to 2 — about 6.6 µs against an 88 µs call. The correctness fix above is the part that matters.

## Tests

`layer_test.py` builds the model above and asserts both the recorded `Node` kwargs and the numeric output, for `training=True`, `False` and `None`. Removing the copy fails all three.

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

**Review by hertschuh — CHANGES_REQUESTED** [review]
This one should wait on and be rebased on https://github.com/keras-team/keras/pull/23288

**Comment by hertschuh** [comment]
Please narrow down the scope of this PR, it includes many other changes.

File: keras/src/layers/layer.py [hunk 1]
```diff
@@ -891,15 +891,38 @@ def maybe_convert(x):
...
+        # `training=<bool-or-None>` shape holds no array-like leaf for
+        # `maybe_convert` to act on.
+        kwargs_may_need_convert = bool(kwargs) and not (
+            len(kwargs) == 1
+            and "training" in kwargs
+            and (
+                kwargs["training"] is True
+                or kwargs["training"] is False
+                or kwargs["training"] is None
+            )
```
Comment (hertschuh) [inline keras/src/layers/layer.py]: Remove this part

File: keras/src/layers/layer.py [hunk 2]
```diff
@@ -891,15 +891,38 @@ def maybe_convert(x):
                 backend.set_keras_mask(y, mask)
             return y
 
+        # `fit()`/`evaluate()`/`predict()` always pass `training=` explicitly,
+        # so a non-empty `kwargs` is not by itself a reason to convert: the
+        # `training=<bool-or-None>` shape holds no array-like leaf for
+        # `maybe_convert` to act on.
```
Comment (hertschuh) [inline keras/src/layers/layer.py]: Shorten this comment. This is sufficient:
```
# `kwargs` is usually a dict with just the `training` argument, which doesn't need conversion.
```

File: keras/src/layers/layer.py [hunk 3]
```diff
@@ -891,15 +891,38 @@ def maybe_convert(x):
...
         if (
-            kwargs
+            kwargs_may_need_convert
             or len(args) != 1
             or not is_backend_tensor_or_symbolic(args[0], allow_none=False)
             or backend.standardize_dtype(args[0].dtype) != self.input_dtype
         ) and self._convert_input_args:
             args = tree.map_structure(maybe_convert, args)
             kwargs = tree.map_structure(maybe_convert, kwargs)
+        elif not kwargs_may_need_convert and kwargs:
```
Comment (hertschuh) [inline keras/src/layers/layer.py]:
```python
elif not kwargs_may_need_convert and kwargs:
```

This seems wrong. if `kwargs_may_need_convert` is True, we necessarily did the block above. And anyway, to be on the safe side, and to minimize cognitive load, we should always create a new dict to avoid issues with mutations, even with an empty dict.

```python
else:
    kwargs = dict(kwargs)
```

File: keras/src/layers/layer.py [hunk 4]
```diff
@@ -891,15 +891,38 @@ def maybe_convert(x):
...
+            # Skipping conversion leaves `kwargs` bound to the very dict that
+            # `original_kwargs` refers to, and later steps mutate it: step 5
+            # populates call-context args, step 6 populates masks, and
+            # `CallSpec` pops call-context args that `call()` does not accept.
+            # Copy so the `Node` recorded in step 8 keeps exactly what the
+            # caller passed. (The conversion branch above already rebinds
+            # `kwargs` to a fresh dict, which is what makes it safe there.)
```
Comment (hertschuh) [inline keras/src/layers/layer.py]: Shorten this comment. This is enough:
```python
# `kwargs` get mutated later, so create a copy.
```

File: keras/src/layers/layer_test.py [hunk 5]
```diff
@@ -2038,3 +2039,117 @@ def call(self, x):
...
+        layer = MaskArg()
+        x = ops.convert_to_tensor(np.random.rand(2, 3).astype("float32"))
+        layer(x)
+
+        self.assertGreater(
+            self._count_map_structure_calls(
+                lambda: layer(x, training=False, mask=None)
+            ),
+            0,
+        )
```
Comment (hertschuh) [inline keras/src/layers/layer_test.py]: Please remove these 3 tests and `_count_map_structure_calls`.

`_count_map_structure_calls` verifies an implementation detail of how the tree API is used, but that doesn't prove the layer code is correct or incorrect.

Instead, we should use an end to end example exercising the actual bug, which is what you did with the last test. So just keep the last test.

**Review by hertschuh — COMMENTED** [review]
Thanks for finding this!

Can you update the issue URL in the description? It's incorrect.

**Comment by hertschuh** [comment]
Oh and then mark it as "Ready for Review".

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!
